handler/src/test/java/io/netty/handler/ssl/OptionalSslHandlerTest.java

```diff
@@ -0,0 +1,99 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import static org.mockito.Mockito.*;
+
+@RunWith(MockitoJUnitRunner.class)
+public class OptionalSslHandlerTest {
+
+    private static final String SSL_HANDLER_NAME = "sslhandler";
+    private static final String HANDLER_NAME = "handler";
+
+    @Mock
+    private ChannelHandlerContext context;
+
+    @Mock
+    private SslContext sslContext;
+
+    @Mock
+    private ChannelPipeline pipeline;
+
+    @Before
+    public void setUp() throws Exception {
+        when(context.pipeline()).thenReturn(pipeline);
+    }
+
+    @Test
+    public void handlerRemoved() throws Exception {
+        OptionalSslHandler handler = new OptionalSslHandler(sslContext, SSL_HANDLER_NAME, HANDLER_NAME);
+        final ByteBuf payload = Unpooled.copiedBuffer("plaintext".getBytes());
+        try {
+            handler.decode(context, payload, null);
+            verify(pipeline).remove(handler);
+        } finally {
+            payload.release();
+        }
+    }
+
+    @Test
+    public void handlerReplaced() throws Exception {
+        final ChannelHandler nonSslHandler = Mockito.mock(ChannelHandler.class);
+        OptionalSslHandler handler = new OptionalSslHandler(sslContext, SSL_HANDLER_NAME, HANDLER_NAME) {
+            @Override
+            protected ChannelHandler newNonSslHandler(ChannelHandlerContext context) {
+                return nonSslHandler;
+            }
+        };
+        final ByteBuf payload = Unpooled.copiedBuffer("plaintext".getBytes());
+        try {
+            handler.decode(context, payload, null);
+            verify(pipeline).replace(handler, HANDLER_NAME, nonSslHandler);
+        } finally {
+            payload.release();
+        }
+    }
+
+    @Test
+    public void sslHandlerReplaced() throws Exception {
+        final SslHandler sslHandler = Mockito.mock(SslHandler.class);
+        OptionalSslHandler handler = new OptionalSslHandler(sslContext, SSL_HANDLER_NAME, HANDLER_NAME) {
+            @Override
+            protected SslHandler newSslHandler(SslContext sslContext, ChannelHandlerContext context) {
+                return sslHandler;
+            }
+        };
+        final ByteBuf payload = Unpooled.wrappedBuffer(new byte[] { 22, 3, 1, 0, 5 });
+        try {
+            handler.decode(context, payload, null);
+            verify(pipeline).replace(handler, SSL_HANDLER_NAME, sslHandler);
+        } finally {
+            payload.release();
+        }
+    }
```

Comment:
nah... I think it will be good enough to check that the handler is still in the pipeline if you pass a buffer with not enough readable bytes into decode(...) as it will return early.